data/blog/improving-nodejs-loader-performance.md

```diff
@@ -0,0 +1,180 @@
+---
+title: 'Improving Node.js loader performance'
+date: '2023-12-12'
+tags: ['node.js', 'performance', 'esm', 'cjs', 'commonjs']
+summary: 'CommonJS and ES modules are 2 sides of a coin. Node.js supports both of them. So, how can we improve the performance of Node.js loaders?'
+images: ['/images/improving-nodejs-loader-performance/hero.jpg']
+layout: PostLayout
+canonicalUrl: improving-nodejs-loader-performance
+authors: ['yagiznizipli']
+---
+
+Node.js supports 2 different modules. EcmaScript and CommonJS modules. ES modules are the official standard for modules in JavaScript and they are supported by all modern browsers. CommonJS modules are the modules that Node.js uses by default. They are not supported by browsers and they are not the official standard. However, they are still widely used.
+
+## How does Node.js load the entry point?
+
+In order to differentiate which loader to use, Node.js depends on several factors. The most important one is the file extension. If the file extension is `.mjs`, Node.js will use the ES module loader. If the file extension is `.cjs`, Node.js will use the CommonJS module loader. If the file extension is `.js`, Node.js will use the CommonJS module loader if the `package.json` file has `"type": "commonjs"` field (or simply doesn't have the `type` field). If the `package.json` file has `"type": "module"` field, Node.js will use the ES module loader.
```

Comment:
really good explanation of how the package.json changes the decision

Reply:
Thanks!